frontend/src/metabase-lib/lib/queries/NativeQuery.ts

```diff
@@ -50,23 +50,25 @@ export const NATIVE_QUERY_TEMPLATE: NativeDatasetQuery = {
 ///////////////////////////
 // QUERY TEXT TAG UTILS
 
-// Matches all snippet, card, and variable template tags. See unit tests for `recognizeTemplateTags` for examples
-const TAG_REGEX: RegExp =
-  /\{\{\s*((snippet:\s*[^}]+)|[A-Za-z0-9_\.]+?|(#[0-9]*(?:-[a-z0-9-]*)?))\s*\}\}/g;
+const VARIABLE_TAG_REGEX: RegExp = /\{\{\s*([A-Za-z0-9_\.]+)\s*\}\}/g;
+const SNIPPET_TAG_REGEX: RegExp = /\{\{\s*(snippet:\s*[^}]+)\s*\}\}/g;
+export const CARD_TAG_REGEX: RegExp =
+  /\{\{\s*(#([0-9]*)(-[a-z0-9-]*)?)\s*\}\}/g;
+const TAG_REGEXES: RegExp[] = [
+  VARIABLE_TAG_REGEX,
+  SNIPPET_TAG_REGEX,
+  CARD_TAG_REGEX,
+];
 
 // look for variable usage in the query (like '{{varname}}').  we only allow alphanumeric characters for the variable name
 // a variable name can optionally end with :start or :end which is not considered part of the actual variable name
 // expected pattern is like mustache templates, so we are looking for something like {{category}}
 // anything that doesn't match our rule is ignored, so {{&foo!}} would simply be ignored
 // See unit tests for examples
 export function recognizeTemplateTags(queryText: string): string[] {
-  const tagNames = [];
-  let match;
-  while ((match = TAG_REGEX.exec(queryText)) != null) {
-    tagNames.push(match[1]);
-  }
-
-  // eliminate any duplicates since it's allowed for a user to reference the same variable multiple times
+  const tagNames = TAG_REGEXES.flatMap(r =>
+    Array.from(queryText.matchAll(r)),
+  ).map(m => m[1]);
```

Comment:
Note this change is behaviour preserving, but I wanted to split up `TAG_REGEX` to reuse the CARD_TAG_REGEX component of it.